### PR TITLE
Unauthorized client side routing

### PR DESCRIPTION
**Pull request by trainerbill**
Reintroducing to master branch.  closing https://github.com/meanjs/mean/pull/727

@ilanbiala @lirantal @rhutchison @codydaig

### CODE REVIEW

File: modules/core/client/config/core.client.routes.js [hunk 1]
```diff
@@ -16,6 +16,14 @@ angular.module('core').config(['$stateProvider', '$urlRouterProvider',
...
         url: '/not-found',
         templateUrl: 'modules/core/views/404.client.view.html'
-      });
+      })
+      .state('bad-request', {
+        url: '/bad-request',
+        templateUrl: 'modules/core/views/400.client.view.html'
+      })
+      .state('unauthorized', {
+				url: '/unauthorized',
```
Comment (ilanbiala) [inline modules/core/client/config/core.client.routes.js]: indentation

File: modules/core/client/views/modal.client.view.html [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<div class="modal-header">
+  <h3 class="modal-title">{{modal.title}}</h3>
+</div>
+<div class="modal-body">
+  {{modal.body}}
+  <div data-ng-if="modal.items">
```
Comment (ilanbiala) [inline modules/core/client/views/modal.client.view.html]: data-

**Comment by trainerbill** [comment]
@ilanbiala Not that I am looking at this, I don't think we should do interceptors for bad request on the client side....   Server side is throwing them quite a bit, even for things like article is not found.  Guessing we don't want a redirect on these errors...  Or do we?

File: modules/core/client/views/modal.client.view.html (hunk 2 above)
Comment (lirantal) [inline modules/core/client/views/modal.client.view.html]: We have another PR for data- handling and I think at the moment we can leave the `data-*` stuff in there as is until we get a better decision on that change.

**Comment by lirantal** [comment]
@trainerbill can you elaborate more on the requirement/purpose of this PR? I'm not really following what's it about.

**Comment by rhutchison** [comment]
> Server side is throwing them quite a bit, even for things like article is not found. Guessing we don't want a redirect on these errors... Or do we?

@trainerbill yes, this is the intention. 404 should go use $state.transitionTo on the client to go to the 404 page.  This will all make better sense when resolves are used properly.

I am not a fan of having the pub/sub implementation here, I don't see the value of doing it this way.  The only thing I can think of is you tried implementing state.go in the http interceptor and ran in to the circular reference issue.  You need to use `$injector` like so, `$injector.get('$state').transitionTo('bad-request');`

**Comment by trainerbill** [comment]
@lirantal HTTP interceptors will do certain actions if the server responds with certain error codes.  So for like an unauthorized 401 response I created a route so that the client will redirect to the route so the user knows they are unauthorized.  This will come in handy when a user views an article then adds /edit to the end of the URL of an article that they don't own.  I suggest pulling the branch and checking it out.  Login with a user account and try to go to /admin/users.  View an article the user doesn't own and add a /edit to the URL and try and edit the article.  You will go to the unauthorized route.

@rhutchison The example you gave me was based on pub/sub.  I actually like the idea.  So then in the module controllers we can broadcast the unauth event and it does the same thing every time for certain events.  $state couldn't be injected into config so that is why I went with pub/sub but after doing it I kinda like it.  As a framework/middleware we should really have events like this to make it easier on the modules.  Just do a $rootScope.broadcast(events) and we handle everything for the modules.

As for 404, the only reason I don't like it is because the backend throws them a lot.  I don't think that redirecting to a bad-request route is a better option then just displaying the error on the client since the bad request error message is variable.  Then the user has to hit the back button to get back to their previous state.

**Comment by trainerbill** [comment]
@rhutchison Removed Modal, Removed Pub/Sub, changed unauthorized to forbidded to align with the http status code.  401 goes to authentication 403 goes to forbidden.

**Comment by rhutchison** [comment]
LGTM - thank you.

**Comment by trainerbill** [comment]
@lirantal.  This is ready to merge

**Comment by lirantal** [comment]
@ilanbiala ?

**Comment by mleanos** [comment]
@trainerbill I pulled this branch and tested. A few things I noticed... Forgive me if you're aware of these issues.. 

When not signed in, I go to an invalid route. I'm redirected to the Bad Request state. Now if sign in successfully, I'm redirected back to the Bad Request state. 

If a 400 status is returned from the server, it redirects to the Bad Request, even if the 400 was caused by an invalid form submission; for instance, incorrect credentials at login.

I saw mention of the issue where the user can add `/edit` to an article they are viewing & get to that state. However, they are correctly sent to the Forbidden state if they attempt to submit the form. What is the plan with addressing that? Wouldn't we want to just avoid allowing that state transition in the first place? Perhaps, that's outside the scope of this PR.

@trainerbill Overall this looks really good. Thanks.

**Comment by trainerbill** [comment]
@mleanos Thanks for taking it for a test drive.  I was actually worried on the bad request 400 interceptor as the backend seems to throw them quite a bit.  I am going to remove the 400 interceptor just to get the PR in.  The forbidden route and interceptor is what I really want in.  @rhutchison any issue with this?

File: modules/core/client/services/interceptors/auth.interceptor.client.service.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+'use strict';
+
+//Articles service used for communicating with the articles REST endpoints
```
Comment (rhutchison) [inline modules/core/client/services/interceptors/auth.interceptor.client.service.js]: remove unrelated comment

File: modules/core/client/app/init.js [hunk 4]
```diff
@@ -4,13 +4,16 @@
 angular.module(ApplicationConfiguration.applicationModuleName, ApplicationConfiguration.applicationModuleVendorDependencies);
 
 // Setting HTML5 Location Mode
-angular.module(ApplicationConfiguration.applicationModuleName).config(['$locationProvider',
-  function ($locationProvider) {
-    $locationProvider.html5Mode(true).hashPrefix('!');
-  }
+angular.module(ApplicationConfiguration.applicationModuleName).config(['$locationProvider', '$httpProvider',
+function ($locationProvider, $httpProvider) {
```
Comment (rhutchison) [inline modules/core/client/app/init.js]: fix indentation

File: modules/core/client/app/init.js [hunk 5]
```diff
@@ -24,11 +27,12 @@ angular.module(ApplicationConfiguration.applicationModuleName).run(function ($ro
...
         event.preventDefault();
-        $state.go('authentication.signin', {}, {
-          notify: false
-        }).then(function () {
-          $rootScope.$broadcast('$stateChangeSuccess', 'authentication.signin', {}, toState, toParams);
-        });
+        if (Authentication.user !== undefined && typeof Authentication.user === 'object') {
+          $state.go('forbidden');
+        }
+        else {
```
Comment (rhutchison) [inline modules/core/client/app/init.js]: should be on previous line

File: modules/core/client/config/core.client.routes.js [hunk 6]
```diff
@@ -2,20 +2,28 @@
 
 // Setting up route
 angular.module('core').config(['$stateProvider', '$urlRouterProvider',
-  function ($stateProvider, $urlRouterProvider) {
+function ($stateProvider, $urlRouterProvider) {
```
Comment (rhutchison) [inline modules/core/client/config/core.client.routes.js]: fix indentation

**Comment by rhutchison** [comment]
LGTM thanks @trainerbill

**Comment by mleanos** [comment]
@trainerbill I see the 400 is removed, so the issue with the invalid form submission is resolved. Thanks. 

I'm still seeing that when I get sent to the not-found state, and then login I'm getting redirected back to the not-found state. Were you waiting to handle that in a separate PR?

Other than that everything looks good to me.

@rhutchison I relinquish my stance on the order of which to merge these PR's. Sounds like we need to discuss, or get feedback, on #807 anyways. 

This LGTM, and appears to be ready for merging.

**Comment by codydaig** [comment]
LGTM

**Comment by rhutchison** [comment]
@mleanos @trainerbill @codydaig  Please review #843 
> I'm still seeing that when I get sent to the not-found state, and then login I'm getting redirected back to the not-found state. Were you waiting to handle that in a separate PR?

**Comment by trainerbill** [comment]
Don't merge yet.  Writing tests

**Comment by trainerbill** [comment]
@lirantal @rhutchison Ready to merge

**Comment by rhutchison** [comment]
LGTM

**Pull request merged**